server/ServerFSM.cpp

```diff
@@ -2167,62 +2164,53 @@ sc::result WaitingForMPGameJoiners::react(const JoinGame& msg) {
-            {
-                collision = false;
-                roles.Clear();
-                if (!server.IsAvailableName(new_player_name) || server.IsAuthRequiredOrFillRoles(new_player_name, roles)) {
-                    collision = true;
-                } else {
-                    for (std::pair<int, PlayerSetupData>& plr : m_lobby_data->m_players) {
-                        if (plr.second.m_empire_name == new_player_name) {
-                            collision = true;
-                            break;
-                        }
+        bool collision = true;
+        std::size_t t = 1;
+        while (collision &&
+               t <= m_lobby_data->m_players.size() + server.Networking().GetCookiesSize() + 1)
+        {
+            collision = false;
+            roles.Clear();
+            if (!server.IsAvailableName(new_player_name) || server.IsAuthRequiredOrFillRoles(new_player_name, roles)) {
+                collision = true;
+            } else {
+                for (std::pair<int, PlayerSetupData>& plr : m_lobby_data->m_players) {
+                    if (plr.second.m_empire_name == new_player_name) {
+                        collision = true;
+                        break;
                     }
                 }
-
-                if (collision)
-                    new_player_name = player_name + std::to_string(++t); // start alternative names from 2
             }
 
-            if (collision) {
-                player_connection->SendMessage(ErrorMessage(str(FlexibleFormat(UserString("ERROR_PLAYER_NAME_ALREADY_USED")) % original_player_name),
-                                                            true));
-                server.Networking().Disconnect(player_connection);
-                return discard_event();
-            }
+            if (collision)
+                new_player_name = player_name + std::to_string(++t); // start alternative names from 2
+        }
 
-            if (!player_connection->HasAuthRole(Networking::ROLE_CLIENT_TYPE_PLAYER)) {
-                player_connection->SendMessage(ErrorMessage(UserStringNop("ERROR_CLIENT_TYPE_NOT_ALLOWED"), true));
-                server.Networking().Disconnect(player_connection);
-                return discard_event();
-            }
+        if (collision) {
+            player_connection->SendMessage(ErrorMessage(str(FlexibleFormat(UserString("ERROR_PLAYER_NAME_ALREADY_USED")) % original_player_name),
+                                                        true));
+            server.Networking().Disconnect(player_connection);
+            return discard_event();
+        }
 
-            fsm.EstablishPlayer(player_connection, new_player_name, client_type, client_version_string, roles);
+        if (!player_connection->HasAuthRole(Networking::ROLE_CLIENT_TYPE_PLAYER)) {
+            player_connection->SendMessage(ErrorMessage(UserStringNop("ERROR_CLIENT_TYPE_NOT_ALLOWED"), true));
+            server.Networking().Disconnect(player_connection);
```

Comment:
Unless a player connection has no roles allowed, could assign a new role instead of disconnecting.

Reply:
Fixed.